astacus/common/rohmustorage.py

```diff
@@ -148,20 +163,22 @@ def _loaded_public_key_lookup(self, key_id: str) -> RSAPublicKey:
         return public_key
 
     @rohmu_error_wrapper
-    def _upload_key_from_file(self, key: str, f: BinaryIO, file_size: int) -> StorageUploadResult:
+    def _upload_key_from_file(self, key: str, f: IO[bytes], file_size: int) -> StorageUploadResult:
         encryption_key_id = self.config.encryption_key_id
         compression = self.config.compression
         metadata = RohmuMetadata()
-        wrapped_file = f
-        if compression.algorithm:
-            metadata.compression_algorithm = compression.algorithm
-            wrapped_file = CompressionStream(wrapped_file, compression.algorithm, compression.level)
+        wrapped_file: IO[bytes] | Stream = f
+        if compression.algorithm_value:
+            metadata.compression_algorithm = RohmuCompressionType(compression.algorithm)
+            wrapped_file = CompressionStream(wrapped_file, compression.algorithm_value, compression.level)
         if encryption_key_id:
             metadata.encryption_key_id = encryption_key_id
             rsa_public_key = self._loaded_public_key_lookup(encryption_key_id)
             wrapped_file = EncryptorStream(wrapped_file, rsa_public_key)
         rohmu_metadata = metadata.dict(exclude_defaults=True, by_alias=True)
-        self.storage.store_file_object(key, wrapped_file, metadata=rohmu_metadata)
+        # BaseTransfer.store_file_object is compatible with file descriptors of type IO[bytes] and Stream
+        # However, the BinaryIO parameter type is too strict for the union type we want to pass here.
+        self.storage.store_file_object(key, wrapped_file, metadata=rohmu_metadata)  # type: ignore[arg-type]
```

Comment:
It's a bit risky to silence type exactly at the boundary to an external library. Especially since we know `store_file_object` can have multiple implementation for each storage providers and where having each party uphold the promise of the interface type would be important for making sure this will work, without testing the product of all library users with all storage providers.

This looks like a bug in rohmu then, it shouldn't take a BinaryIO and use a type that only requires the "read" part of the interface ?

(That's not blocking, since the type was completely unchecked before this PR, so it's not worse than before.)

Reply:
Rohmu's storage interface needs to be fixed. As far as I can tell, the `fd` should abide by a protocol that implements `seek`, `tell` and `read`. Not sure how it type checks at the moment, since `Stream` is not a `BinaryIO` subclass.